libr/core/dmh_glibc.inc.c

```diff
@@ -100,32 +100,231 @@ static GHT GH(get_main_arena_with_symbol)(RCore *core, RDebugMap *map) {
+			size = section->size;
+			break;
+		}
+	}
+
+	RBinFile *libc_bf = r_bin_cur (bin);
+	r_bin_file_delete (bin, libc_bf->id);
+	r_bin_file_set_cur_binfile (bin, bf);
+	return size;
+}
+
+static ut8 *GH(get_section_bytes)(RCore *core, const char *path, const char *section_name) {
+	RBin *bin = core->bin;
+	RBinFile *bf = r_bin_cur (bin);
+	bool found_section = false;
+	GHT paddr, size;
+
+	RBinFileOptions opt;
+	r_bin_file_options_init (&opt, -1, 0, 0, false);
+	if (!r_bin_open (bin, path, &opt)) {
+		R_LOG_ERROR ("get_section_bytes: r_bin_open failed on path %s", path);
+		return NULL;
+	}
+
+	RBinFile *libc_bf = r_bin_cur (bin);
+	RList *sections = r_bin_get_sections (bin);
+	RBinSection *section;
+	RListIter *iter;
+	ut8 *buf = NULL;
+	r_list_foreach (sections, iter, section) {
+		if (!strcmp (section->name, section_name)) {
+			found_section = true;
+			paddr = section->paddr;
+			size = section->size;
+			break;
+		}
+	}
+
+	if (!found_section) {
+		R_LOG_WARN ("get_section_bytes: section %s not found", section_name);
+		goto cleanup_exit;
+	}
+
+	// eprintf ("get_section_bytes: section found: %s size: %#08x  paddr: %#08x\n", section_name, size, paddr);
+	buf = calloc (size, 1);
+	if (!buf) {
+		R_LOG_ERROR ("get_section_bytes: calloc failed");
+		goto cleanup_exit;
+	}
+
+	st64 read_size = r_buf_read_at (libc_bf->buf, paddr, buf, size);
+
+	if (read_size != size) {
+		R_LOG_ERROR ("get_section_bytes: section read unexpected size: %#08x  (section->size: %d)", read_size, size);
+		free (buf);
+		buf = NULL;
+	}
+
+cleanup_exit:
+	r_bin_file_delete (bin, libc_bf->id);
+	r_bin_file_set_cur_binfile (bin, bf);
+	return buf;
+}
+
+R_API double GH(get_glibc_version)(RCore *core, const char *libc_path) {
+	double version = 0.0;
+
+	// First see if there is a "__libc_version" symbol
+	// If yes read version from there
+	GHT version_symbol = GH (get_va_symbol) (core, libc_path, "__libc_version");
+	if (version_symbol != GHT_MAX) {
+		FILE *libc_file = fopen (libc_path, "rb");
+		if (libc_file == NULL) {
+			R_LOG_WARN ("resolve_glibc_version: Failed to open %s", libc_path);
+			return false;
+		}
+		// TODO: futureproof this
+		char version_buffer[5] = {0};
+		fseek (libc_file, version_symbol, SEEK_SET);
+		if (fread (version_buffer, 1, 4, libc_file) != 4)	{
+			R_LOG_WARN ("resolve_glibc_version: Failed to read 4 bytes of version symbol");
+			return false;
+		};
+
+		fclose (libc_file);
+		if (!r_regex_match ("\\d.\\d\\d", "e", version_buffer)) {
+			R_LOG_WARN ("resolve_glibc_version: Unexpected version format: %s", version_buffer);
+			return false;
+		}
+		version = strtod (version_buffer, NULL);
+		R_LOG_INFO ("libc version %.2f identified from symbol", version);
+		return version;
+	}
+
+	// Next up we try to read version from banner in .rodata section
+	// also inspired by pwndbg
+	const ut8 *rodata_bytes = GH (get_section_bytes) (core, libc_path, ".rodata");
+	GHT rodata_size = GH (get_section_size) (core, libc_path, ".rodata");
```

Comment:
Why dont get bytes and size in the same call instead of iterating twice?

Reply:
Makes sense in my mind to have `get_section_bytes` and `get_section_size` since you might want to get only the size of a section and you need the bytes

Reply:
These apis are used only in your code and you are calling both, this implies doing the work twice, which makes its slower, and also you can just have one api with the last argument as size_t *size to return the size of the section itself.